### ISSUE TITLE
Setting up rtags for iOS/Objective-C development

### ISSUE THREAD

**Issue by jojojames** [Andersbakken/rtags#738]
Hi, I'm having trouble setting up rtags to work with an iOS (built from xcode project).

I've set up a dummy project to test rtags with.
https://github.com/jojojames/DummyXcode

I also detailed the steps I took here.
https://raw.githubusercontent.com/jojojames/DummyXcode/master/setup.org

Relevant compile_commands.json is here.
https://github.com/jojojames/DummyXcode/blob/master/build/reports/compile_commands.json

Error:
$ rc -J build/reports/
** Error
[100%] 3/3 11:42:22 ~/Developer/DummyXcode/DummyXcode/AppDelegate.m in 8984ms. (35 syms, 24 symNames, 1 includes, 1 of 2 files, symbols: 46 of 47, 414 cursors, 1 queried 0ms) (8982/0/1ms). (priority 0)
Jobs took 9.00s, (avg 1.50s). We're using 0mb of memory.
Output from /Users/james/Developer/DummyXcode/DummyXcode/AppDelegate.m:
2431 errors generated.
2432 errors generated.
error: unknown argument: '-fembed-bitcode-marker'
/Users/james/Developer/DummyXcode/DummyXcode/AppDelegate.h:9:9: fatal error: could not build module 'UIKit'

clang installed using,
$ brew install llvm --with-libcxx --with-clang --without-assertions --with-rtti

I'm guessing it might have something to do with incompatibility between xcode's llvm/clang and the one installed by brew.

https://stackoverflow.com/questions/32595742/homebrew-llvm-build-cannot-find-ios-simulator-library-containing-wordexp-symbol

Not sure how to get this working but next steps may be to compile llvm from source with iOS flags.

It seems to be working (rtags-find-symbol-at-point) for the rest of the project but not for the UIKit namespace which is consistent with the error message.

Another datapoint, using some other completion libraries that also has a clang backing. Passing in a few compiler flags is enough to get (UIKit) completions working.
[
'-isysroot',
'/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk',
'-I/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/7.3.0/include'
]

Any help here would be appreciated and thanks for the great package!

**Comment by Andersbakken** [comment]
Hi

Thanks for a great bug report.
I blacklisted a couple of these arguments that causes problems for rtags and I think it should work better now. Note that our obj-c support is a little lacking currently. I believe we don't handle CXCursor_ObjCMessageExpr correctly.

**Comment by jojojames** [comment]
Thanks.

What's the effect of the blacklist?

Resulting log after pulling and building rtags again.

**Comment by Andersbakken** [comment]
The blacklist basically means that we won't pass this argument on to libclang. I can't reproduce the problem with UIKit though. It seems my clang finds UIKit here:

/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/UIKit.framework/Headers/UIKit.h

I'm not entirely sure what doesn't work in terms of objective-c(++) but I know we had some special cased code at some point that got lost in a refactor. It's quite possible that most/all things work. I'm not sure we have that many users of rtags for objective-c(++). 

I built clang myself I think. I used the cmake build system and can't quite figure out what options (if any) I passed.

**Comment by jojojames** [comment]
Thanks! It's working like a charm right now (in the sample project as well as a largish project).

Some notes (would also be happy to split some of these into separate issues).
* company-rtags doesn't seem to start automatically for semantic completions. It seems like it only works from manual completions, though it seems like it does do keyword completions.

 - (void)viewViewAppear:(BOOL)animated is a method from UIKit that is returned from rdm (checked company-diag) and does show up if completion is initiated manually.

* company-rtags post completion for methods in Obj-c doesn't wrap the ( ).
- (void)viewDidAppear:(BOOL)animated
Completion becomes - (void)viewDidAppear:(**BOOL**) where the bold is the text selection.

* cmake didn't seem to pull my clang/llvm environment variables properly. I ended up setting each of the cmake variables manually. I tried debugging this for a bit but ended up just going for the quick and dirty approach.

//
    message(STATUS "$ENV{LIBCLANG_LIBDIR}") <------ Works
    message(STATUS "$LIBCLANG_LIBDIR") <------ Always empty even if exported before calling make
//
    execute_process(COMMAND ${LIBCLANG_LLVM_CONFIG_EXECUTABLE} --libdir OUTPUT_VARIABLE LIBCLANG_LIBDIR OUTPUT_STRIP_TRAILING_WHITESPACE)
    message(STATUS ${LIBCLANG_LIBDIR}) <---- LIBCLANG_LIBDIR is always empty here though calling llvm-config --libdir directly in terminal always returns the right path.

* I added a setup script here for those wondering into this issue page.
https://github.com/jojojames/.emacs.d/blob/master/tools/rtags_install.sh
